### PR TITLE
make journal pluggable; record deals, sealing, wdpost, mempool events

### PR DESCRIPTION
**Pull request by raulk**
___Please merge with a squash___

---

Addresses part of #2441; the other part (in-memory journal implementation) is hosted in Project Oni: https://github.com/filecoin-project/oni/pull/226.

# What has been done here

## New journal events

* Sealing state transition events are now recorded in the journal.
  * Required a change in storage-fsm to emit notifications.
* Deal state transitions are now recorded in the journal.
* Windowed PoSt events (run, abort, faults_processed, recoveries_processed, proofs_processed) are now recorded in the journal.
* Mempool events (add, remove, repub) are now recorded in the journal. Add and remove are disabled by default.

## Journal framework

* Make the Journal _not global_; instead, use `fx` to inject it where needed.
* Support typed journal events.
* By default, Lotus constructs the filesystem-based journal, but it can be overridden with another implementation through a constructor option.
* Ability to suppress/disable certain types of events, such as those known to be noisy and not relevant during production use, but important for testing (e.g. mempool add/drop).
* In order to avoid runtime costs, suppressed event types need to be disabled during construction, by overriding the `DisabledEvents` fx key.
* By default, we suppress the noisy mempool events. Oni tests will pass in an empty disabled set, so we get access to everything.
* Sugar API function `MaybeRecordEvent` that lazily creates the event only if the EventType is enabled.

# Example journal events

## Sealing state machine transitions

```json
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:05:00.556940387Z","Data":{"SectorNumber":19,"SectorType":2,"From":"WaitSeed","After":"Committing","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:07:46.797337768Z","Data":{"SectorNumber":19,"SectorType":2,"From":"Committing","After":"CommitWait","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:11:59.258008229Z","Data":{"SectorNumber":19,"SectorType":2,"From":"CommitWait","After":"FinalizeSector","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:11:59.655317986Z","Data":{"SectorNumber":19,"SectorType":2,"From":"FinalizeSector","After":"Proving","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:44:00.581866465Z","Data":{"SectorNumber":20,"SectorType":2,"From":"WaitDeals","After":"Packing","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:44:13.940864008Z","Data":{"SectorNumber":20,"SectorType":2,"From":"Packing","After":"PreCommit1","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:45:03.216173326Z","Data":{"SectorNumber":20,"SectorType":2,"From":"PreCommit1","After":"PreCommit2","Error":""}}
{"System":"storage","Event":"sealing_states","Timestamp":"2020-08-11T15:46:30.365776445Z","Data":{"SectorNumber":0,"SectorType":0,"From":"","After":"Empty","Error":""}}
```

## Windowed PoSt

```json
{"System":"storage","Event":"wdpost","Timestamp":"2020-08-11T15:20:30.401238096Z","Data":{"State":"abort","Deadline":{"CurrentEpoch":10099,"PeriodStart":8419,"Index":28,"Open":10099,"Close":10159,"Challenge":10079,"FaultCutoff":10029},"Height":10158,"TipSet":[{"/":"bafy2bzacecv7qpj2o4i3wqiv2sia6sbuwiicequvvfwofybhownpdxjjll5d4"}]}}
{"System":"storage","Event":"wdpost","Timestamp":"2020-08-11T15:20:30.401491228Z","Data":{"State":"started","Deadline":{"CurrentEpoch":10159,"PeriodStart":8419,"Index":29,"Open":10159,"Close":10219,"Challenge":10139,"FaultCutoff":10089},"Height":10158,"TipSet":[{"/":"bafy2bzacecv7qpj2o4i3wqiv2sia6sbuwiicequvvfwofybhownpdxjjll5d4"}]}}
{"System":"storage","Event":"wdpost","Timestamp":"2020-08-11T15:20:30.403296622Z","Data":{"State":"recoveries_processed","Deadline":{"CurrentEpoch":10159,"PeriodStart":8419,"Index":29,"Open":10159,"Close":10219,"Challenge":10139,"FaultCutoff":10089},"Height":10158,"TipSet":[{"/":"bafy2bzacecv7qpj2o4i3wqiv2sia6sbuwiicequvvfwofybhownpdxjjll5d4"}],"Recoveries":{"Declarations":[],"MessageCID":null}}}
{"System":"storage","Event":"wdpost","Timestamp":"2020-08-11T15:20:30.403301022Z","Data":{"State":"faults_processed","Deadline":{"CurrentEpoch":10159,"PeriodStart":8419,"Index":29,"Open":10159,"Close":10219,"Challenge":10139,"FaultCutoff":10089},"Height":10158,"TipSet":[{"/":"bafy2bzacecv7qpj2o4i3wqiv2sia6sbuwiicequvvfwofybhownpdxjjll5d4"}],"Faults":{"Declarations":[],"MessageCID":null}}}
```

## mpool

Add and remove events are disabled by default, but this is how add events look.

```json
{"System":"mpool","Event":"add","Timestamp":"2020-08-11T15:35:42.640242628Z","Data":{"Action":"add","Messages":[{"Version":0,"To":"t01464","From":"t3q2sc3zvdhcwhuttajyv6zhlfmqfffan736fc552zu5m6vofneecmsbs5b3clrpgexklaqn75fgztim5y6rgq","Nonce":12673,"Value":"3727422619627069401","GasLimit":21651646,"GasFeeCap":"1064652776302","GasPremium":"4","Method":6,"Params":"igMZ1EjYKlgpAAGC4gOB6AIge/eO8WleRP1CBG7vb588+wOuByEifkZXavzVsjicJDEZH7uAGgAIu6P0AAAA","CID":{"/":"bafy2bzacebq3qfiynqc5oopqwmlhljsleybimkofuek5efthmfa7ang2hfa6i"}}]}}
{"System":"mpool","Event":"add","Timestamp":"2020-08-11T15:35:43.220132993Z","Data":{"Action":"add","Messages":[{"Version":0,"To":"t05","From":"t3qtgrsjee4tubfpzf2hjdmeqonkoz4c3llracv2lnwtanmkrexbtvcln7zvjwmptdblq67lkle66npbbnvbnq","Nonce":3351,"Value":"0","GasLimit":60881214,"GasFeeCap":"1088471905894","GasPremium":"2","Method":4,"Params":"gYGCi9gqWCgAAYHiA5IgIA/Jg0iXqR68K/HsuKblVjnlXjJzGmfzDFZ8XfYX5Hg4GgAEAAD0WDEDhM0ZJITk6BK/JdHSNhIOap2eC2tcQCrpbbTA1iokuGdRLb/NU2Y+Ywrh761LJ7zXQwC0DnhkewoJInBjaWRzIjogWwoJCXsKCQkJIi8iOiAiYmFmazJiemFjZWM1aWRuempxZHNubjM0ZmVnZnAzNGhlaHhyeXBmbmFmbHAycGxzeWplNGM1a2N3YnhsY3EiCgkJfQoJXQp9ChkzDRoACZ4ARQBQd12AQEBYYQKorCx9eeMPROqSS3EBVOkpB+D8ED3F9BFhdy/zLuENKoM5wpVzsczB3BqebR+LWIcScjNM4bDlT///WH/qlvkMTD5ac6E3gz1FzjmK9EhipOCZD0Qz146C1fmYq4DDNbg=","CID":{"/":"bafy2bzacebxx7hk7t6wpmehvo7ociw7tadntksjfwdyjnniygtqmulh7pamqo"}}]}}
{"System":"mpool","Event":"add","Timestamp":"2020-08-11T15:35:44.044902337Z","Data":{"Action":"add","Messages":[{"Version":0,"To":"t01045","From":"t3spbjekuheshmqj6gsz3fbrzozejq5aiyntqtaxuhptnemc5k4bolfgtyc76efrkwj4j6nz5vhlhf765i3fda","Nonce":26757,"Value":"3727361587939049590","GasLimit":14942648,"GasFeeCap":"335131587179","GasPremium":"2","Method":6,"Params":"igMZM23YKlgpAAGC4gOB6AIgf5Uw+fvcv1mECsio0Zoaqh44mYmGhagTyLH3TVIhzxUZH6mAGgAX0aL0AAAA","CID":{"/":"bafy2bzacecprq5z43bfucpg2hwh2faak7pjicz7jb6k74vaskqio5qjr5p5o6"}}]}}
{"System":"mpool","Event":"add","Timestamp":"2020-08-11T15:35:44.319591507Z","Data":{"Action":"add","Messages":[{"Version":0,"To":"t02220","From":"t3snngyn6annziz5heowxfkkd6jcpn5bizr2xkgptxxbatq2ttbuo56hl2xwnxsuqv3uj7nn6nzrxh73ufisha","Nonce":34002,"Value":"3727361587939049590","GasLimit":23323009,"GasFeeCap":"1","GasPremium":"1","Method":6,"Params":"igMaAASgqNgqWCkAAYLiA4HoAiCbrcfJNcDgfR5x8EXLAxkbPKhz0e+O427yQ0dFq8VQGhkiEYAaABfan/QAAAA=","CID":{"/":"bafy2bzacedzl7tmdqbvdct7ionck3ct5pibnxrrvt2cfocyvi7qcr65pqlarq"}}]}
```

## Deals

```json
{"System":"markets:storage:provider","Event":"deal_complete","Timestamp":"2020-08-11T15:43:00.600472498Z","Data":{"ID":18281,"Deal":{"Proposal":{"PieceCID":{"/":"baga6ea4seaqemba7nfrn6r2rv7x7q5uaaqrzdhyux3jpa2vvas4qqu7x2fae4hq"},"PieceSize":131072,"VerifiedDeal":false,"Client":"t3ulfykd6qimw63fbk6lcdjqnw6b7jqxwt2h7omwfdrawbssizutamjfax5g6iiqbkx22pt2yl2dtgk7vnj67a","Provider":"t03212","Label":"{\n\t\"pcids\": [\n\t\t{\n\t\t\t\"/\": \"bafk2bzacedmlz6vhpcemizgcax5almemh5o2f76vnnyk3r3dsofehi375rre2\"\n\t\t}\n\t]\n}\n","StartEpoch":13075,"EndEpoch":593058,"StoragePricePerEpoch":"2000000","ProviderCollateral":"0","ClientCollateral":"0"},"ClientSignature":{"Type":0,"Data":null},"ProposalCid":{"/":"bafyreicm35wkw6vrzbtpf3mj4zcquhi4h7qs6cmrfxatqagqcrqnm6hngy"},"AddFundsCid":null,"PublishCid":null,"Miner":"","Client":"12D3KooWJU6qqDH4DJtYfZrqiqb2dyNtgW2HNjhPmr6whSU6MgNb","State":4,"PiecePath":"","MetadataPath":"","SlashEpoch":0,"FastRetrieval":true,"Message":"","StoreID":null,"FundsReserved":"\u003cnil\u003e","Ref":{"TransferType":"graphsync","Root":{"/":"bafk2bzacedmlz6vhpcemizgcax5almemh5o2f76vnnyk3r3dsofehi375rre2"},"PieceCid":null,"PieceSize":0},"AvailableForRetrieval":false,"DealID":18281},"State":{"SectorStartEpoch":0,"LastUpdatedEpoch":0,"SlashEpoch":0},"Height":0}}
```

### CODE REVIEW

File: journal/memory.go
Comment (Kubuxu) [inline journal/memory.go:112]: `len(include)`, no reason to make it static

File: journal/types.go
Comment (yusefnapora) [inline journal/types.go:25]: Why not just have `AddEntry` ignore disabled events?

Comment (raulk) [reply journal/types.go:25]: Because (a) that would require a map lookup if implemented naively, and (b) by that time, the caller has already constructed the data object.

I think a cleaner, more ergonomic approach would be to replace this with a `Journal#GetEventType() EventType` method, and add a `EventType#Enabled() bool`. So you'd lose the extra stray boolean, and the state would be self-contained in the event type.

Comment (raulk) [reply journal/types.go]: @yusefnapora take a look at the refactor here: https://github.com/filecoin-project/lotus/pull/2455/commits/bef7d64e666242c19e4f76cb88422c64947e8f00

File: go.mod
Comment (raulk) [inline go.mod:40]: This points to a PR branch: https://github.com/filecoin-project/storage-fsm/pull/54. Needs to be updated when the PR is merged.

File: journal/fs.go
Comment (raulk) [inline journal/fs.go:1]: Most of this was copied as-is.

File: node/modules/storageminer.go
Comment (raulk) [inline node/modules/storageminer.go:146]: Changed this to an `fx.In` wrapped parameter style. The constructor was getting rather bloated with inline args.

File: storage/wdpost_sched.go
Comment (raulk) [inline storage/wdpost_sched.go:35]: These could actually be splintered into separate event types.

File: markets/storageadapter/provider.go
Comment (raulk) [inline markets/storageadapter/provider.go:51]: For performance reasons, an array of `journal.EventType`s is what I settled on as the lowest overhead, nicely encapsulated technique to initialize and store `EventTypes` during construction time, and index into them during runtime.

**Comment by lanzafame** [comment]
@raulk How costly is having the mpool journal turned on? As this would be extremely useful for tracking the mpool from an operations standpoint.

**Comment by magik6k** [comment]
I don't think we want to journal everything getting in/out of mempool by default, but it's reasonable for addresses we control

**Comment by raulk** [comment]
@magik6k this patch affects a bunch of files across the board, and it's already conflicting. Can we review and merge this quickly with restrictive settings (i.e. no mpool add/remove by default), and iterate on `next`? Follow-up changes we might want to make:

* journal mpool add and remove for addresses we control. (i.e. segregate add and rm into mpool/{add,rm}:{ours,other})
* add a configuration setting for disabled events, rather can have these hardcoded, so that folks like @lanzafame can tune to their liking.

**Comment by raulk** [comment]
@magik6k @Kubuxu can I get a review of the (now conflicting) changes here? I'll merge from next once there's feedback.

File: storage/miner.go
Comment (magik6k) [inline storage/miner.go:54]: Ideally this would live inside the `sealing` package. We want to move `sealing.New` construction to DI, but it's somewhat entangled with the `Miner` struct here. We should try to not make this worse

File: storage/wdpost_run.go
Comment (magik6k) [inline storage/wdpost_run.go:158]: Should probably change `sm, err := s.api.MpoolPushMessage` below to set with `=`

File: storage/wdpost_run.go
Comment (magik6k) [inline storage/wdpost_run.go:265]: Same here

File: storage/wdpost_run.go
Comment (magik6k) [inline storage/wdpost_run.go:534]: And here

File: storage/wdpost_sched.go
Comment (magik6k) [reply storage/wdpost_sched.go:37]: Probably worth doing, window post journal will likely be the one of the most looked at

**Review by magik6k — COMMENTED** [review]
Sorry for taking so long on this.

Looks pretty good, just a couple of comments

File: journal/nil.go
Comment (raulk) [reply journal/nil.go:14]: I don't think so. This journal is designed for testing or for running stuff outside the lotus daemon (e.g. CLI tools). Things _will_ call it all the time; this journal just no-ops. Maybe we can scream loudly when registering events, but I think it adds no value.

File: node/builder.go
Comment (raulk) [reply node/builder.go:126]: Very good point! Right now this wasn't being noticed because the other invokes don't require the journal, but I'll reorder things.

File: node/modules/storageminer.go
Comment (raulk) [reply node/modules/storageminer.go:171]: Yep, some methods are getting very loaded.

File: storage/wdpost_run.go
Comment (raulk) [reply storage/wdpost_run.go:158]: Oops, this was a merge error, thanks for catching!

Comment (raulk) [reply storage/wdpost_run.go:158]: Actually, the `sm, err := s.api.MpoolPushMessage` below _does_ set on this variable. It is what the spec calls a "redeclaration": https://golang.org/ref/spec#Short_variable_declarations. But I will move this logic to `runPost`, and make `checkNextRecoveries` and `checkNextFaults` return the data we need. That way we can also log if an error happens.

File: storage/wdpost_sched.go
Comment (raulk) [reply storage/wdpost_sched.go:37]: Makes sense, will do.

Comment (raulk) [reply storage/wdpost_sched.go]: Done.

**Comment by raulk** [comment]
Note that I don’t have a miner, so I haven’t tested this to the degree I'd be satisfied. Please test with a miner of yours before merging! There may be hidden bugs that weren’t caught by tests.

**Review by whyrusleeping — APPROVED** [review]
lets doooo ittttt

**Comment by RobQuistNL** [comment]
This is freaking awesome <3

**Pull request merged**
